### PR TITLE
Symfony events

### PR DESCRIPTION
**Pull request by hlohaus**
###   1. Why is this change necessary?

Alternative to #2143. We use symfony as event listener list.

### 2. What does this change do, exactly?
- You can now use symfony listener and subscribers.
- Adds symfony console events
- You can use enlight events in symfony and symfony events in enlight.
- HttpKernel added. So KernelEvents and smyfony controller are useable.

Breaks:
- Return values of the methods getListeners and getAllListeners in event manger changed from Enlight_Event_Handler to callable.
- Renamed getListener to getMethod in Enlight_Event_Handler_Plugin. getListener returns now a callable.

Example with symfony event subscriber:
 https://github.com/hlohaus/SheTest/blob/master/Subscribers/ConsoleEvents.php#L30

Example controller:

https://github.com/hlohaus/SheTest/blob/master/SheTest.php#L49

2x faster... 💃 

### 6. Checklist

- [x] I have written tests and verified that they fail without my change
- [x] I have squashed any insignificant commits
- [x] This change has comments for package types, values, functions, and non-obvious lines of code
- [x] I have read the contribution requirements and fulfil them.

### CODE REVIEW

File: engine/Shopware/Kernel.php [hunk 1]
```diff
@@ -54,6 +54,7 @@
 use Symfony\Component\DependencyInjection\ContainerBuilder;
 use Symfony\Component\DependencyInjection\Loader\XmlFileLoader;
 use Symfony\Component\DependencyInjection\ParameterBag\ParameterBag;
+use Symfony\Component\EventDispatcher\DependencyInjection\RegisterListenersPass;
```
Comment (shyim) [inline engine/Shopware/Kernel.php:56]: Please add a new require to the composer.json, this works currently cause an other dependency require event-dispatcher

File: composer.json [hunk 2]
```diff
@@ -81,7 +81,8 @@
         "league/flysystem-aws-s3-v3": "1.0.19",
         "superbalist/flysystem-google-storage": "6.0.0",
         "stecman/symfony-console-completion": "0.10.1",
-        "beberlei/doctrineextensions": "1.2.1"
+        "beberlei/doctrineextensions": "1.2.1",
+        "symfony/event-dispatcher": "^4.3"
```
Comment (shyim) [inline composer.json]: sort it to the other symfony packages 😅

File: engine/Shopware/Components/Controller.php [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+    private $container;
+
+    public function __construct(ContainerInterface $container)
+    {
+        $this->container = $container;
+    }
+
+    public function dispatch(Request $request): Response
+    {
+        if (!$request instanceof \Enlight_Controller_Request_RequestHttp) {
```
Comment (hlohaus) [inline engine/Shopware/Components/Controller.php]: Maybe we cab change the request object in shopware.php. Then it will not be converted. @shyim

File: build/build.xml [hunk 4]
```diff
@@ -238,11 +238,15 @@
         <exec executable="${script.phpunit}" failonerror="true" dir="${test.dir.shopware}">
             <arg value="--config=phpunit_unit.xml.dist"/>
             <arg value="--exclude-group=elasticSearch"/>
+            <arg value="--stop-on-failure"/>
```
Comment (shyim) [inline build/build.xml]: Can you revert this changes

File: composer.json [hunk 5]
```diff
@@ -51,6 +51,7 @@
         "symfony/options-resolver": "3.4.29",
         "symfony/expression-language": "3.4.29",
         "symfony/web-link": "3.4.29",
+        "symfony/event-dispatcher": "^4.3",
```
Comment (shyim) [inline composer.json]: Please use 3.4.29 like the other packages from symfony

**Comment by PascalThesing** [comment]
@hlohaus have you tested the different hooks?

**Comment by hlohaus** [comment]
I have linked in description a Plugins with Symfony Events. Or wich hooks do you krank?

**Comment by PascalThesing** [comment]
I am thinking about Shopware hooks. They use the same syntax e.g. 'Shopware_Controllers_Frontend_Checkout::confirmAction::after'

**Comment by hlohaus** [comment]
Yes, they should work. There are tests

**Comment by shyim** [comment]
Hey,
we discussed your PR. Can you split the Events part and Httpkernel part into own Pull Requests?
Currently we are not sure how many Symfony we want to integrate into Shopware 5, so we can discuss that part later on.
We currently don't want two dispatch systems and two ways to do the same thing.
